lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (BasicBlock &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      MetaAddress Start = getBasicBlockPC(&BB);
+      model::BasicBlock Block{ Start };
+      Block.End = getFinalAddressOfBasicBlock(&BB);
+
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto LastInstructionAddress = getFinalAddressOfBasicBlock(Current);
+        if (LastInstructionAddress > Block.End)
+          Block.End = LastInstructionAddress;
+
+        for (BasicBlock *Succ : successors(Current)) {
+          if (GCBI::isJumpTarget(Succ)) {
+            MetaAddress Destination = getBasicBlockPC(Succ);
+            auto Edge = makeEdge(Destination,
+                                 model::FunctionEdgeType::DirectBranch);
+            Block.Successors.insert(Edge);
+          } else if (UnexpectedPCCloned == Succ && succ_size(Current) == 1) {
+            // Need to create an edge only when `unexpectedpc` is the unique
+            // successor of the current basic block.
+            Block.Successors.insert(makeEdge(MetaAddress::invalid(),
+                                             model::FunctionEdgeType::LongJmp));
+          } else {
+            Instruction *I = &(*Succ->begin());
+
+            if (auto *Call = getCallTo(I, PreHookMarker)) {
+              MetaAddress Destination;
+              model::FunctionEdgeType::Values Type;
+              auto *CalleePC = Call->getArgOperand(1);
+
+              // Direct or indirect call?
+              if (isa<ConstantStruct>(CalleePC)) {
+                Destination = MetaAddress::fromConstant(CalleePC);
+                Type = model::FunctionEdgeType::FunctionCall;
+              } else {
+                Destination = MetaAddress::invalid();
+                Type = model::FunctionEdgeType::IndirectCall;
+              }
+
+              auto Edge = makeEdge(Destination, Type);
+              auto DestTy = Oracle.getFunctionType(Destination);
+              if (DestTy == FunctionTypeValue::NoReturn)
+                cast<model::CallEdge>(Edge.get())->IsNoReturn = true;
+
+              Block.Successors.insert(Edge);
+            } else if (auto *Call = getCallTo(I, "function_call")) {
+              // Marker can be erased now, no longer needed.
+              Call->eraseFromParent();
+
+              auto Destination = getBasicBlockPC(Succ);
+              auto Edge = makeEdge(Destination,
+                                   model::FunctionEdgeType::FakeFunctionCall);
+              Block.Successors.insert(Edge);
+            } else {
+              Queue.insert(Succ);
+            }
+          }
+        }
+      }
+
+      CFG.insert(Block);
+    }
+  }
+
+  return CFG;
+}
+
+template<class FO>
+FunctionSummary
+CFEPAnalyzer<FO>::analyze(ArrayRef<GlobalVariable *> ABIRegisters,
```

Comment:
`CFEPAnalyzer` has access to `GCBI`, why pass in `ABIRegisters`?
Having a single argument of `analyze` is very elegant!